### PR TITLE
WIP: Relates to #81. Update to Electron v3.1.0

### PR DESCRIPTION
**Pull request by ltfschoen**
* Update to Electron v3.1.0 - https://github.com/electron/electron/releases

* Update Readme

* Update dependencies versions

* Add .nvmrc

### CODE REVIEW

File: .nvmrc [hunk 1]
```diff
@@ -0,0 +1 @@
+10.11.0
```
Comment (amaurymartiny) [inline .nvmrc]: Why is this needed actually? Is there a package we use that require at least this version of node? Is it redundant with what you added in package.json?

File: README.md [hunk 2]
```diff
@@ -31,6 +31,66 @@ And below are the lower-level packages, used internally, or by advanced users.
...
+Install at least `yarn` version 1.4.2 and [Node.js >=10.11.0](https://nodejs.org/en/)
+
+```
+yarn --version // Should be at least 1.4.2
+```
+
+#### Tests
+
+```
+export NPM_TOKEN=''; yarn test
```
Comment (amaurymartiny) [inline README.md]: No need for `export NPM_TOKEN='';`. Can you confirm @ltfschoen?

File: README.md [hunk 3]
```diff
@@ -31,6 +31,66 @@ And below are the lower-level packages, used internally, or by advanced users.
...
+```
+
+#### Build
+
+```
+export NPM_TOKEN=''; yarn build
```
Comment (amaurymartiny) [inline README.md]: No need for export NPM_TOKEN='';

File: README.md [hunk 4]
```diff
@@ -31,6 +31,66 @@ And below are the lower-level packages, used internally, or by advanced users.
...
+4. Create a branch and update any dependencies that it indicates are out of date.
+
+```
+git checkout -b <INSERT_YOUR_BRANCH_NAME>
+```
+
+5. Run tests, linting, and build
+
+```
+export NPM_TOKEN=''; yarn test; yarn lint; yarn build
```
Comment (amaurymartiny) [inline README.md]: No need for export NPM_TOKEN='';

File: package.json [hunk 5]
```diff
@@ -1,6 +1,7 @@
 {
   "name": "js-libs",
   "description": "A collection of JavaScript libraries for dapp development.",
+  "version": "v3.0.26",
```
Comment (amaurymartiny) [inline package.json]: No need to have version on the root package.json, see for example [babel](https://github.com/babel/babel/blob/master/package.json)

File: packages/abi/package.json [hunk 6]
```diff
@@ -17,11 +17,12 @@
   "scripts": {
     "docs": "typedoc",
     "prebuild": "rimraf lib",
-    "build": "tsc",
+    "build": "yarn && ./node_modules/.bin/tsc",
```
Comment (amaurymartiny) [inline packages/abi/package.json]: On my machine there's no need for this addition. Did you get an error?

File: packages/electron/package.json [hunk 7]
```diff
@@ -17,25 +17,26 @@
...
+    "electron-dl": "^1.12.0",
     "promise-any": "^0.2.0",
-    "sntp": "^3.0.1"
+    "sntp": "^3.0.2"
   },
   "devDependencies": {
     "@types/async-retry": "^1.2.1",
     "@types/checksum": "^0.1.30",
-    "electron": "^2.0.2"
+    "electron": "^3.1.0",
```
Comment (amaurymartiny) [inline packages/electron/package.json]: I want this package to be available for all electron versions. The only weak link is `electron-dl`, which I am not 100% sure is compatible with Electron 3-4, but let's assume yes until proven wrong (and do some testing on Fether). All other packages are only nodejs.

So replace with "^2.0.2 | ^3.1.0 | ^4.0.1"

File: packages/electron/package.json [hunk 8]
```diff
@@ -17,25 +17,26 @@
...
+    "typescript": "^3.1.6"
   },
   "peerDependencies": {
-    "electron": "^2.0.3"
+    "electron": "^3.1.0"
```
Comment (amaurymartiny) [inline packages/electron/package.json]: "^2.0.2 | ^3.1.0 | ^4.0.1"

File: packages/light.js-react/package.json [hunk 9]
```diff
@@ -38,7 +38,8 @@
     "enzyme-adapter-react-16": "^1.2.0",
     "jest-enzyme": "^6.0.3",
     "react": "^16.4.2",
-    "react-dom": "^16.4.2"
+    "react-dom": "^16.4.2",
+    "typescript": "^3.1.6"
```
Comment (amaurymartiny) [inline packages/light.js-react/package.json]: devDep

File: packages/light.js/package.json [hunk 10]
```diff
@@ -37,7 +37,8 @@
     "@types/debug": "^0.0.30",
     "@types/memoizee": "^0.4.2",
     "eventemitter3": "^3.1.0",
-    "rxjs": "~6.2.2"
+    "rxjs": "~6.2.2",
+    "typescript": "^3.1.6"
```
Comment (amaurymartiny) [inline packages/light.js/package.json]: devDep

File: README.md [hunk 11]
```diff
@@ -31,6 +31,66 @@ And below are the lower-level packages, used internally, or by advanced users.
...
+
+2. Clone your fork
+```
+git clone https://github.com/<INSERT_YOUR_GITHUB_USERNAME>/js-libs
+```
+
+3. Check outdated dependencies
+
+```
+npm outdated
```
Comment (amaurymartiny) [inline README.md]: yarn outdated

File: README.md [hunk 12]
```diff
@@ -31,6 +31,66 @@ And below are the lower-level packages, used internally, or by advanced users.
...
+```
+
+4. Create a branch and update any dependencies that it indicates are out of date.
```
Comment (amaurymartiny) [inline README.md]: I'd prefer that it'd be us who update dependencies. They sometimes make things break. Or even better: implement #28.

File: .nvmrc (hunk 1 above)
Comment (ltfschoen) [reply .nvmrc]: In the Electron [Release Notes for v4.0.0](https://github.com/electron/electron/releases/tag/v4.0.0) it says: 

> Bumped minimum supported macOS version to 10.10.

However the release notes for v4.0.1 or v3.1.0 don't mention a minimum version requirement https://github.com/electron/electron/releases

File: README.md (hunk 2 above)
Comment (ltfschoen) [reply README.md]: Yes confirmed. It is no longer required

File: package.json (hunk 5 above)
Comment (ltfschoen) [reply package.json]: If I pushed my changes in this branch 'luke-81-electron-3' of js-libs repo without a version number in its package.json, then when I tried to test using this branch 'luke-309-electron-3' of Fether on this line of code https://github.com/paritytech/fether/pull/349/files#diff-52ba44750a2113719866b7d98b31af30R42 `"@parity/electron": "https://github.com/paritytech/js-libs.git#luke-81-electron-3",`, then when I tried to build Fether with `yarn; yarn start` it would give me an error that it required a version number in the package.json file

File: packages/abi/package.json (hunk 6 above)
Comment (ltfschoen) [reply packages/abi/package.json]: Yes, I was getting the following errors:
```
@parity/abi: /bin/sh: tsc: command not found
@parity/electron: /bin/sh: tsc: command not found 
```
In the details of the commit message I included the log https://github.com/paritytech/js-libs/pull/90/commits/73c2042822fa558d4f29a3fcbf4151bdc278db3f

File: package.json (hunk 5 above)
Comment (ltfschoen) [reply package.json]: It produced the following:
```
error Can't add "@parity/electron": invalid package version undefined.
```

In the details of the commit message I included the full log https://github.com/paritytech/js-libs/pull/90/commits/8b347e047c260e53f3da0955bf87bb2c10de2744

File: packages/light.js-react/package.json (hunk 9 above)
Comment (ltfschoen) [reply packages/light.js-react/package.json]: It doesn't look like any change is required because `"typescript": "^3.1.6"` is already within `"devDependencies": {`

File: packages/electron/package.json (hunk 7 above)
Comment (ltfschoen) [reply packages/electron/package.json]: If I also update Fether's fether-electron package's package.json with:
```
"devDependencies": {
    ...
    "electron": "^2.0.2 | ^3.1.0 | ^4.0.1",
```

Then when I run `yarn; yarn start` it prompts me to choose the version, but it shows versions prior as well like 1.x.x
```
? Please choose a version of "electron" from this list: (Use arrow keys)
❯ 4.0.1 
  4.0.0 
  4.0.0-nightly.20181010 
...
  4.0.0-beta.1 
  3.1.0 
  3.1.0-beta.5 
...
  3.0.4 
(Move up and down to reveal more choices)
```

File: packages/light.js-react/package.json (hunk 9 above)
Comment (amaurymartiny) [reply packages/light.js-react/package.json]: screenshot 2019-01-11 at 10 04 21

My bad! I read a bit too quickly

File: README.md [hunk 13]
```diff
@@ -31,6 +31,66 @@ And below are the lower-level packages, used internally, or by advanced users.
...
+
+5. Run tests, linting, and build
+
+```
+yarn test; yarn lint; yarn build
+```
+
+6. Push the branch to your fork of the repo
+
+7. Integrate the updated library as a dependency. Example: If you are using the 'electron' package in another project, then update the package.json file to temporarily use your branch instead of the public NPM registry version
```
Comment (amaurymartiny) [inline README.md]: @ltfschoen I understand now where your errors come from. I would like to remove this step 7: For `@parity/*` packages, I would advise against doing `yarn add github:something` into your own project.

The reason is:
- right now we develop in `src/`
- the CI builds, outputs in `lib/`, and pushes `lib/` to npm
- when you `yarn add @parity/*` it downloads `lib/` from npm, and the main file is `lib/index.js`

However, when you `yarn add github:something`, it does:
- clone the repo
- run yarn build inside the cloned folder
- hopefully, if the build doesn't error, it generates a `lib/` folder 

Jaco and I used to try to do this method above, however there's always some troubles here and there in the "hopefully" step. That's why I prefer to remove it. Some alternative solutions:
- (ideal solution) wait for the npm package to be published. If you need to test a package, it's better to add a test directly in this repo, so that the test is self-contained.
- use npm link
- do the copy/paste dance:
```
# in js-libs/
yarn build
cp -r packages/my-package/lib ../fether/node_modules/@parity/my-package/lib
# in fether
yarn start
```

File: .nvmrc [hunk 14]
```diff
@@ -0,0 +1 @@
+10.10.0
```
Comment (amaurymartiny) [inline .nvmrc]: My questions are:
- If your nvm has v11, does it change to v10.10.0? (not necessary)
- Is it enough to only put `"engine"` in package.json?

File: packages/abi/package.json (hunk 6 above)
Comment (amaurymartiny) [reply packages/abi/package.json]: These errors are in the fether, right? If yes, then see my comment about removing step 7

File: packages/electron/package.json (hunk 7 above)
Comment (amaurymartiny) [reply packages/electron/package.json]: In js-libs, we put "^2.0.2 | ^3.1.0 | ^4.0.1", which means "this library supports these versions of Electron."

In Fether, we choose one specific version of Electron (4.0.1). So in fether it should only be "^4.0.1"

File: .nvmrc (hunk 14 above)
Comment (ltfschoen) [reply .nvmrc]: You're right. I'll remove the .nvmrc file. It's more useless than I thought

If I'm using the wrong Node.js version it warns me when I run `yarn` as follows:
```
scon @ ~/code/src/paritytech/js-libs - [luke-81-electron-4] $ nvm use 10.6.0
Now using node v10.6.0 (npm v6.1.0)
scon @ ~/code/src/paritytech/js-libs - [luke-81-electron-4] $ yarn
yarn install v1.12.3
[1/5] 🔍  Validating package.json...
error js-libs@3.0.26: The engine "node" is incompatible with this module. Expected version ">=10.10.0". Got "10.6.0"
error Found incompatible module
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
```

You're right, there's just as many steps whether we do or don't have the .nvmrc file. The important thing is to have the "engine" specified.

```
scon @ ~/code/src/paritytech/js-libs - [luke-81-electron-4] $ node --version
v10.6.0
scon @ ~/code/src/paritytech/js-libs - [luke-81-electron-4] $ nvm use
No .nvmrc file found
```

With .nvmrc file:
```
scon @ ~/code/src/paritytech/js-libs - [luke-81-electron-4] $ node --version
v10.6.0
scon @ ~/code/src/paritytech/js-libs - [luke-81-electron-4] $ nvm use
Found '/Users/scon/code/src/paritytech/js-libs/.nvmrc' with version <10.10.0>
N/A: version "10.10.0 -> N/A" is not yet installed.

You need to run "nvm install 10.10.0" to install it before using it.
scon @ ~/code/src/paritytech/js-libs - [luke-81-electron-4] $
```

**Comment by ltfschoen** [comment]
Closing as replaced by PR https://github.com/paritytech/js-libs/pull/91 and comments being incorporated.

File: README.md (hunk 2 above)
Comment (ltfschoen) [reply README.md]: I just created another issue. It looks like this issue came back again
